pkg/azuredisk/device_perf_optimization_linux.go

```diff
@@ -0,0 +1,296 @@
+		readAheadKb)
+
+	err = echoToFile(maxSectorsKb, filepath.Join("/sys/block", deviceName, "queue/max_sectors_kb"))
+	if err != nil {
+		klog.Errorf("OptimizeDevicePerformance: Could not set max_sectors_kb for device %s. Error: %v", deviceName, err)
+		return err
+	}
+
+	err = echoToFile(scheduler, filepath.Join("/sys/block", deviceName, "queue/scheduler"))
+	if err != nil {
+		klog.Errorf("OptimizeDevicePerformance: Could not set scheduler for device %s. Error: %v", deviceName, err)
+		return err
+	}
+
+	err = echoToFile("1", filepath.Join("/sys/block", deviceName, "queue/iosched/fifo_batch"))
+	if err != nil {
+		klog.Errorf("OptimizeDevicePerformance: Could not set scheduler for device %s. Error: %v", deviceName, err)
+		return err
+	}
+
+	err = echoToFile("1", filepath.Join("/sys/block", deviceName, "queue/iosched/writes_starved"))
+	if err != nil {
+		klog.Errorf("OptimizeDevicePerformance: Could not set scheduler for device %s. Error: %v", deviceName, err)
+		return err
+	}
+
+	err = echoToFile(queueDepth, filepath.Join("/sys/block", deviceName, "device/queue_depth"))
+	if err != nil {
+		klog.Errorf("OptimizeDevicePerformance: Could not set queue_depth for device %s. Error: %v", deviceName, err)
+		return err
+	}
+
+	err = echoToFile(nrRequests, filepath.Join("/sys/block", deviceName, "queue/nr_requests"))
+	if err != nil {
+		klog.Errorf("OptimizeDevicePerformance: Could not set nr_requests for device %s. Error: %v", deviceName, err)
+		return err
+	}
+
+	err = echoToFile(readAheadKb, filepath.Join("/sys/block", deviceName, "queue/read_ahead_kb"))
+	if err != nil {
+		klog.Errorf("OptimizeDevicePerformance: Could not set max_sectors_kb for device %s. Error: %v", deviceName, err)
+		return err
+	}
+
+	err = echoToFile("0", filepath.Join("/sys/block", deviceName, "queue/wbt_lat_usec"))
+	if err != nil {
+		klog.Errorf("OptimizeDevicePerformance: Could not set max_sectors_kb for device %s. Error: %v", deviceName, err)
+		return err
+	}
+
+	err = echoToFile("0", filepath.Join("/sys/block", deviceName, "queue/rotational"))
+	if err != nil {
+		klog.Errorf("OptimizeDevicePerformance: Could not set max_sectors_kb for device %s. Error: %v", deviceName, err)
+		return err
+	}
+
+	return err
+}
+
+// getDeviceName gets the device name from the device lunpath
+// Device lun path is of the format /dev/disk/azure/scsi1/lun0
+func getDeviceName(lunPath string) (deviceName string, err error) {
+	devicePath, err := filepath.EvalSymlinks(lunPath)
+	if err != nil {
+		klog.Errorf("Path %s is not a symlink. Error: %v", lunPath, err)
+		return "", err
+	}
+
+	return filepath.Base(devicePath), nil
+}
+
+// echoToFile echos setting value to the file
+func echoToFile(content string, filePath string) (err error) {
+	klog.V(2).Infof("Echoing %s to %s", content, filePath)
+	cmd := exec.Command("echo", content)
+	outfile, err := os.Create(filePath)
+	if err != nil {
+		return err
+	}
+	defer outfile.Close()
+
+	cmd.Stdout = outfile
+
+	err = cmd.Start()
```

Comment:
echoToFile is similar to what Exec.Command is doing. Difference being we echo to a file vs returning the output.

May be I am not clear with what you mean here. Can you elaborate?

Reply:
would you use ioutil package to open & write file?

Reply:
No

Reply:
Why not append to the file directly rather than spawn a process to echo to a redirected stdout?

[Append Example](https://pkg.go.dev/os@go1.16.4#example-OpenFile-Append)

Reply:
Hi @edreed One of these settings file exhibit some special behavior, echo results in toggling of this setting. Notice how we toggle between none and mq-deadline.
root@abhib-dev-ubuntu-18:/home/azureuser/go/src/sigs.k8s.io/azuredisk-csi-driver# echo mq-deadline > /sys/block/sdc/queue/scheduler
root@abhib-dev-ubuntu-18:/home/azureuser/go/src/sigs.k8s.io/azuredisk-csi-driver# cat /sys/block/sdc/queue/scheduler
[mq-deadline] none
root@abhib-dev-ubuntu-18:/home/azureuser/go/src/sigs.k8s.io/azuredisk-csi-driver# echo none > /sys/block/sdc/queue/scheduler
root@abhib-dev-ubuntu-18:/home/azureuser/go/src/sigs.k8s.io/azuredisk-csi-driver# cat /sys/block/sdc/queue/scheduler
[none] mq-deadline 
root@abhib-dev-ubuntu-18:/home/azureuser/go/src/sigs.k8s.io/azuredisk-csi-driver# ls /sys/block/sdc/queue/scheduler
/sys/block/sdc/queue/scheduler
root@abhib-dev-ubuntu-18:/home/azureuser/go/src/sigs.k8s.io/azuredisk-csi-driver# ls -l /sys/block/sdc/queue/scheduler
-rw-r--r-- 1 root root 4096 May 24 20:28 /sys/block/sdc/queue/scheduler

I did try this io* operations but I was overwriting in what I did. I am not sure why the special behavior with this file. But I can always circle back and make use of IO* operations, when I understand this FS behavior better. For now, I've just made all of it consistent with echo calls.